main.py

```diff
@@ -1,132 +1,379 @@
     # endRow – last row to load, or the last of file, the minimum
-    # cols – columns to load, there may be columns we would not want to load as they are useless
+    # cols – columns to load, there may be columns we would not want to load
+    # as they are useless
     # data – return the loaded partial file
+    try:
+        return load_data_type(path, startRows, endRows, cols, sheet)
+    except:
+        return pd.DataFrame()
+
+
+def save_file_xls2(new_path, temp, sheet):
+    if not os.path.isfile(new_path):
+        pd.DataFrame().to_excel(new_path, sheet_name = sheet)
+    with pd.ExcelWriter(new_path, mode = 'a', if_sheet_exists = 'replace') \
+            as writer:
+        temp.to_excel(writer, sheet_name = sheet)
+
+
+def save_file_xls(path, temp, sheet, newDirectory):
+    new_path = newDirectory+path[-(len(path)-path.rfind("/")):]
+    save_file_xls2(new_path, temp, sheet)
+
+
+def save_file_csv(path, temp, newDirectory):
+    new_path = newDirectory+path[-(len(path)-path.rfind("/")):]
+    temp.to_csv(new_path, mode = 'a')
+
+
+def save_file_type(path, temp, newDirectory, sheet = None):
     if path[-5:] == ".xlsx" or path[-4:] == ".xls":
-        temp = pd.read_excel(path,usecols=cols,nrows=endrows-startrows,skiprows=startrows)
+        save_file_xls(path, temp, sheet, newDirectory)
     elif path[-4:] == ".csv":
-        temp=pd.read_csv(path,usecols=cols,nrows=endrows-startrows,skiprows=startrows)
-    return temp
+        save_file_csv(path, temp, newDirectory)
+
 
-def save_file(path, pred, startrows=None, endrows=None):
-    # save_data(path, startRow, endRow)
-    # path – path to the file to load to,
+def save_file(path, prediction, newDirectory, startRows = None,
+              endRows = None, sheet = 'Sheet1'):
+    # save_data(path, startRows, endRow)
+    # path – path to the file to load to, 
     #               or create a blank file, titled “path_new.csv/xls”
-    # startRow – first row to save
+    # startRows – first row to save
     # endRow – last row to load, or the last of file, the minimum
-    # TODO:find better way to partially save file,
-    #  dont want to take too much memory, or time
-    if path[-5:] == ".xlsx":
-        temp = pd.read_excel(path, nrows=endrows - startrows, skiprows=startrows)
-        new_path = path[:-5] + "_new" + path[-5:]
-        temp = pd.concat(temp, pred)
-        temp.to_excel(new_path)
-    elif path[-4:] == ".xls":
-        temp = pd.read_excel(path, nrows=endrows - startrows, skiprows=startrows)
-        new_path = path[:-4] + "_new" + path[-4:]
-        temp = pd.concat(temp, pred)
-        temp.to_excel(new_path)
-    elif path[-4:] == ".csv":
-        temp = pd.read_csv(path, nrows=endrows - startrows, skiprows=startrows)
-        new_path = path[:-4] + "_new" + path[-4:]
-        temp = pd.concat(temp, pred)
-        temp.to_csv(new_path)
+    temp = load_data(path, startRows, endRows, sheet = sheet)
+    temp["grouping"] = prediction.tolist()
```

Comment:
Not sure what's going on here. Why do you need this? This code is very hard to read. I understand that some of it has to do with the specific library you use to handle the excel / csv. But you need to think about a way to make it easier and clearer. 

Also - use constants for everything that's constant (e.g. `"grouping"`)

Reply:
im re-loading the data, joining it with the predictions then saving it.
i could just save a vector of all the predictions, but i think it would be more useful with all the data and calling that column "grouping"
i have to re-load the data because the data i use gets preprocessed
i could preprocess a copy of the data and lose the re-loading, but it would mean i have to use more memory (exactly what im trying to avoid)